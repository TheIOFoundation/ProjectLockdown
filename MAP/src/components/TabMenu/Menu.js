import { Translation } from 'react-i18next';
import { Ticker } from '../Ticker/Ticker';
import { Expandable } from '../Expandable/Expandable';
import { Settings } from '../Settings/Settings';
import '../../style/main.css';
import './TabMenu.scss';
const Menu = (
    isDark,
    setDarkMode,
    menuItem,
    callback,
    currentDropdown,
    onDropDown,
    onLocateChange,
    locale = {
      t: s => {
        switch (s) {
          case 'menu.userPreferenceSection.theme.action':
            return 'Toggle ';
          case 'menu.userPreferenceSection.theme.light':
            return 'Light mode';
          case 'menu.userPreferenceSection.theme.dark':
            return 'Dark mode';
          default:
            return s;
        }
      },
    }
  ) => {
    switch (menuItem) {
      case 'info':
        return {
          title: 'info',
          template: (
            <>
            <Translation>
  { (t, { 
    i18n
   }) => 
              <>
              <h1>Project Lockdown</h1>
              <p className="ld-alpha">
                {t('menu.informationSection.banner')}
              </p>
              <p>
                <b>{t('menu.informationSection.main.name')}</b>
                {t('menu.informationSection.main.text')}
              </p>
  
              <Expandable
                toggle={t('menu.informationSection.about.title')}
                currentDropdown={currentDropdown}
                onDropDown={onDropDown}
                detail={
                  <p>
                    {t('menu.informationSection.about.text.p1')}
                    <b>{t('menu.informationSection.main.name')}</b>
                    {t('menu.informationSection.about.text.p2')}
                    <b>{t('menu.informationSection.main.name')}</b>
                    {t('menu.informationSection.about.text.p3')}
                  </p>
                }
              />
  
              <Expandable
                toggle={t('menu.informationSection.sources.title')}
                currentDropdown={currentDropdown}
                onDropDown={onDropDown}
                detail={
                  <>
                    <p>
                      <b>{t('menu.informationSection.main.name')}</b>
                      {t('menu.informationSection.sources.subtitle')}
                    </p>
                    <ul className="ld-sources">
                      <li>
                        <a
                          className="ld-link"
                          target="_blank"
                          rel="noopener noreferrer"
                          href="https://TIOF.Click/LockdownData"
                        >
                          {' '}
                          {t(
                            'menu.informationSection.sources.linksList.first.linkTitle'
                          )}
                        </a>
                        <br />
                        {t(
                          'menu.informationSection.sources.linksList.first.text'
                        )}
                      </li>
                      <li>
                        <a
                          className="ld-link"
                          target="_blank"
                          rel="noopener noreferrer"
                          href="https://covid19api.com/"
                        >
                          {t(
                            'menu.informationSection.sources.linksList.second.linkTitle'
                          )}
                        </a>
                        <br />
                        {t(
                          'menu.informationSection.sources.linksList.second.text'
                        )}{' '}
                        <a
                          href="https://systems.jhu.edu/"
                          rel="noopener noreferrer"
                          target="_blank"
                        >
                          {t(
                            'menu.informationSection.sources.linksList.second.highlight'
                          )}
                        </a>
                        <br />
                      </li>
                    </ul>
                    <p>
          {t('menu.informationSection.sources.issues.text')}{' '}
                      <a
                        rel="noopener noreferrer"
                        target="_blank"
                        href="https://github.com/Code-for-All/lockdown/issues"
                      >
                        {t(
                          'menu.informationSection.sources.issues.highlight'
                        )}
                      </a>
                      .
                    </p>
                    <div className="link-container">
                      <a
                        className="ld-link"
                        target="_blank"
                        rel="noopener noreferrer"
                        href="https://tiof.click/LockdownData"
                      >
                        {t(
                          `menu.contribution.contributionLinks.firstLink`
                        )}
                      </a>
                      <a
                        className="ld-link"
                        target="_blank"
                        rel="noopener noreferrer"
                        href="https://tiof.click/pldpsf"
                      >
                        {t(
                          `menu.contribution.contributionLinks.secondLink`
                        )}
                      </a>
                    </div>
                  </>
                }
              />
              <Expandable
                toggle={t('menu.informationSection.credits.title')}
                currentDropdown={currentDropdown}
                onDropDown={onDropDown}
                detail={
                  <p>
                    <b>{t('menu.informationSection.main.name')}</b>
                    {t('menu.informationSection.credits.text')}
                    <a
                      href="https://docs.google.com/spreadsheets/d/1mVyQxxLxAF3E1dw870WHXTOLgYzmumojvzIekpgvLV0/edit?ts=5e74ac83#gid=634415797"
                      target="_blank"
                      rel="noopener noreferrer"
                    >
                      {t('menu.informationSection.credits.highlight')}
                    </a>
                    .
                  </p>
                }
              />
  
              <Expandable
                toggle={t('menu.informationSection.dataPrivacity.title')}
                currentDropdown={currentDropdown}
                onDropDown={onDropDown}
                detail={
                  <>
                    <p>
                      {t(
                        'menu.informationSection.dataPrivacity.paragraphs.p1'
                      )}
                    </p>
                    <p>
                      {t(
                        'menu.informationSection.dataPrivacity.paragraphs.p2'
                      )}
                    </p>
                  </>
                }
              />
              </>
              }
              </Translation>
            </>
          ),
        };
      case 'settings':
        return {
          title: 'settings',
          template: (
            <Settings
              darkMode={isDark}
              setDarkMode={setDarkMode}
              onClose={callback}
              onLocateChange={onLocateChange}
              locale={locale}
            />
          ),
        };
      case 'contribute':
        return {
          title: 'contribute',
          template: (
            <>
              <p>
                Do you have any NPI source you would want to share with us? Please
                follow this link:
              </p>
              <a
                className="ld-link"
                target="_blank"
                rel="noopener noreferrer"
                href="https://docs.google.com/forms/d/e/1FAIpQLSfDWe2qlzUnd3e-YYspMzT9adUswDEYIdJMb7jz7ule34-yiA/viewform"
              >
                Submit data
              </a>
              <p>
                Additionally, you can become part of the project by joining one of
                the following teams:
                <ul>
                  <li>Editors</li>
                  <li>Devs</li>
                  <li>UI/UX</li>
                </ul>
              </p>
              <p>
                Contact us via Telegram at:
                <br />
                <a
                  className="ld-link"
                  target="_blank"
                  rel="noopener noreferrer"
                  href="https://t.me/ProjectLockdown"
                >
                  {' '}
                  @ProjectLockdown
                </a>
              </p>
            </>
          ),
        };
  
      case 'updates':
        return {
          title: 'updates',
          template: <Ticker />,
        };
  
      default:
        return {
          title: '',
          template: <></>,
        };
    }
  };
  
  export default Menu;