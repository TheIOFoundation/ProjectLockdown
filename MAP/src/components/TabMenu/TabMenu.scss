@import '../../style/main.css';

.menu-overlay {
  position: absolute;
  top: 0px;
  left: 0px;
  z-index: 1100;
  background-color: rgba(40, 40, 40, 0.8);
  width: 100%;
  height: 100%;
  @media (max-width: 960px) {
    display: none;
  }
}
.menu-close-btn {
  position: absolute !important;
  color: white;
  background-color: transparent;
  padding: 0px !important;
  top: 90px;
  width: 39px !important;
  height: 39px !important;
  right: 0;
  left: 0;
  margin-right: auto !important;
  margin-left: auto !important;
  border-radius: 50% !important;
  /*box-shadow: 0px 4px 15px rgba(0, 0, 0, 0.3);*/
  display: flex;
  align-items: center;
  justify-content: center;
  &:hover {
    cursor: pointer;
  }
  @media (max-width: 960px) {
    background: transparent !important;
    box-shadow: none;
    border-radius: 0px;
  }
}
.dark .menu-close-btn {
  position: absolute !important;
  color: white;
  background-color: transparent;
  padding: 0px;
  top: 10%;
  width: 39px;
  right: 0;
  left: 0;
  margin-right: auto;
  margin-left: auto;
  border-radius: 50%;
  /*box-shadow: 0px 4px 15px rgba(0, 0, 0, 0.3);*/
  &:hover {
    cursor: pointer;
  }
  @media (max-width: 960px) {
    background: transparent !important;
    box-shadow: none;
    border-radius: 0px;
  }
}
.dark .menu-side-btn {
  background-color: transparent;
  color: white;
  background-image: url(../../assets/icons/menu/black.svg);
  & svg {
    color: white;
  }
  &:hover {
    cursor: pointer;
  }
}
.menu-side-btn {
  position: absolute;
  height: 108px;
  width: 50px;
  z-index: 2000;
  bottom: 150px;
  right: -5px;
  border-radius: 100% 0% 0% 100% / 21% 0% 0% 21%;
  border: 0px;
  background-color: transparent;
  background-image: url(../../assets/icons/menu/white.svg);
  &:focus:not(.leaflet-container) {
    box-shadow: none !important;
  }
  & svg {
    color: #333333;
  }
  &:hover {
    cursor: pointer;
  }
}

@keyframes fadeInRight {
  from {
    opacity: 0;
    width: 0%;
  }

  to {
    opacity: 1;
    width: 8%;
  }
}
@keyframes fadeInRightM {
  from {
    opacity: 0;
    width: 0%;
  }

  to {
    opacity: 1;
    width: 23%;
  }
}
.ld-menu {
  @media (max-width: 960px) {
    width: 56px;
    top: 0px;
    height: 100%;
    animation: fadeInRightM 1s;
  }
  height: 100%;
  @media (min-width: 960px) {
    animation: fadeInRight 1s;
  }
  top: 0;
  bottom: 0;
  width: 56px;
  max-width: 56px;
  position: absolute;
  padding: 0;
  right: 0px;
  background-color: #f2f2f2;
  z-index: 1100;
  .dark & {
    background-color: var(--ld-gray-2) !important;
  }
  & .link-container {
    display: flex;
    justify-content: space-between;
    padding: 25px 0px;
  }
  & .ld-menu-nav {
    display: flex;
    height: 100%;
    width: 100%;
    & nav {
      width: 100%;
      padding: 10px 0px 10px 0px;
      display: flex;
      flex-direction: column;
      justify-content: center;
      & ul {
        padding: 0;
        justify-content: flex-start;
        align-items: left;
        min-height: 50%;
        display: flex;
        flex-direction: column;
        & li {
          margin-bottom: 29%;
          &.hide {
            display: none; /*Just for now*/
          }
          & button {
            background-color: transparent;
          }
          & .ld-menu--active {
            &:focus:not(.leaflet-container) {
              box-shadow: none !important;
            }
            position: relative;
            background-color: transparent;
            .dark & {
              background-color: transparent !important;
              color: white !important;
            }
            & svg {
              .dark & {
                color: #f2f2f2;
              }
              color: #828282;
            }
            &::before {
              position: absolute;
              left: -1px;
              top: 7px;
              height: 15px;
              width: 10px;
              z-index: -1;
              content: ' ';
              border-radius: 0% 100% 100% 69% / 0% 50% 51% 0%;
              border: 0px;
              /*box-shadow: 0px 4px 4px rgba(0, 0, 0, 0.25);*/
              background-color: transparent;
              z-index: 2000;
              border-width: 10px 0 10px 10px;
              border-style: solid;
              border-color: transparent transparent transparent #ffffff;
              .dark & {
                background-color: transparent !important;
                border-color: transparent transparent transparent #333333;
              }
            }
          }
          & svg {
            color: #bdbdbd;
            .dark & {
              color: #bdbdbd;
            }
          }
        }
      }
    }
  }
}

.ld-menu--content {
  display: block;
  top: 0px;
  right: 100%;
  height: 100%;
  position: absolute;
  background-color: white;
  width: 700px;
  border-radius: 0px 0px 0px 0px;
  -moz-border-radius: 0px 0px 0px 0px;
  -webkit-border-radius: 0px 0px 0px 0px;
  overflow: auto;
  box-shadow: 4px 4px 15px rgba(0, 0, 0, 0.25);
  @media (max-width: 960px) {
    border-radius: 0px;
    width: calc(100vw - 100%);
    left: calc(100% - 100vw);
  }
  .dark & {
    background-color: #333333 !important;
  }
}
