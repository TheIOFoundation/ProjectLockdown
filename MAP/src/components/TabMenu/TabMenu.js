import { Component } from 'react';
import '../../style/main.css';
import './TabMenu.scss';
import { Expandable } from '../Expandable/Expandable';
import { Ticker } from '../Ticker/Ticker';
import { Settings } from '../Settings/Settings';
import { installMediaQueryWatcher } from '../../utils/media-query';
import Tabs from '../Tabs/Tabs';
import { burger, close as closeIcon } from '../../assets/icons/icons.js';

const renderMenu = (
  menuItem,
  callback,
  currentDropdown,
  onDropDown,
  onLocateChange,
  locale
) => {
  switch (menuItem) {
    case 'info':
      return {
        title: 'info',
        template: (
          <>
            <h1>Project Lockdown</h1>
            <p className="ld-alpha">
              {locale.t('menu.informationSection.banner')}
            </p>
            <p>
              <b>{locale.t('menu.informationSection.main.name')}</b>
              {locale.t('menu.informationSection.main.text')}
            </p>

            <Expandable
              toggle={locale.t('menu.informationSection.about.title')}
              currentDropdown={currentDropdown}
              onDropDown={onDropDown}
              detail={
                <p>
                  {locale.t('menu.informationSection.about.text.p1')}
                  <b>{locale.t('menu.informationSection.main.name')}</b>
                  {locale.t('menu.informationSection.about.text.p2')}
                  <b>{locale.t('menu.informationSection.main.name')}</b>
                  {locale.t('menu.informationSection.about.text.p3')}
                </p>
              }
            />

            <Expandable
              toggle={locale.t('menu.informationSection.sources.title')}
              currentDropdown={currentDropdown}
              onDropDown={onDropDown}
              detail={
                <>
                  <p>
                    <b>{locale.t('menu.informationSection.main.name')}</b>
                    {locale.t('menu.informationSection.sources.subtitle')}
                  </p>
                  <ul className="ld-sources">
                    <li>
                      <a
                        className="ld-link"
                        target="_blank"
                        rel="noopener noreferrer"
                        href="https://TIOF.Click/LockdownData"
                      >
                        {' '}
                        {locale.t(
                          'menu.informationSection.sources.linksList.first.linkTitle'
                        )}
                      </a>
                      <br />
                      {locale.t(
                        'menu.informationSection.sources.linksList.first.text'
                      )}
                    </li>
                    <li>
                      <a
                        className="ld-link"
                        target="_blank"
                        rel="noopener noreferrer"
                        href="https://covid19api.com/"
                      >
                        {locale.t(
                          'menu.informationSection.sources.linksList.second.linkTitle'
                        )}
                      </a>
                      <br />
                      {locale.t(
                        'menu.informationSection.sources.linksList.second.text'
                      )}{' '}
                      <a
                        href="https://systems.jhu.edu/"
                        rel="noopener noreferrer"
                        target="_blank"
                      >
                        {locale.t(
                          'menu.informationSection.sources.linksList.second.highlight'
                        )}
                      </a>
                      <br />
                    </li>
                  </ul>
                  <p>
                    {locale.t('menu.informationSection.sources.issues.text')}{' '}
                    <a
                      rel="noopener noreferrer"
                      target="_blank"
                      href="https://github.com/Code-for-All/lockdown/issues"
                    >
                      {locale.t(
                        'menu.informationSection.sources.issues.highlight'
                      )}
                    </a>
                    .
                  </p>
                  <div className="link-container">
                    <a
                      className="ld-link"
                      target="_blank"
                      rel="noopener noreferrer"
                      href="https://tiof.click/LockdownData"
                    >
                      {locale.t(
                        `menu.contribution.contributionLinks.firstLink`
                      )}
                    </a>
                    <a
                      className="ld-link"
                      target="_blank"
                      rel="noopener noreferrer"
                      href="https://tiof.click/pldpsf"
                    >
                      {locale.t(
                        `menu.contribution.contributionLinks.secondLink`
                      )}
                    </a>
                  </div>
                </>
              }
            />
            <Expandable
              toggle={locale.t('menu.informationSection.credits.title')}
              currentDropdown={currentDropdown}
              onDropDown={onDropDown}
              detail={
                <p>
                  <b>{locale.t('menu.informationSection.main.name')}</b>
                  {locale.t('menu.informationSection.credits.text')}
                  <a
                    href="https://docs.google.com/spreadsheets/d/1mVyQxxLxAF3E1dw870WHXTOLgYzmumojvzIekpgvLV0/edit?ts=5e74ac83#gid=634415797"
                    target="_blank"
                    rel="noopener noreferrer"
                  >
                    {locale.t('menu.informationSection.credits.highlight')}
                  </a>
                  .
                </p>
              }
            />

            <Expandable
              toggle={locale.t('menu.informationSection.dataPrivacity.title')}
              currentDropdown={currentDropdown}
              onDropDown={onDropDown}
              detail={
                <>
                  <p>
                    {locale.t(
                      'menu.informationSection.dataPrivacity.paragraphs.p1'
                    )}
                  </p>
                  <p>
                    {locale.t(
                      'menu.informationSection.dataPrivacity.paragraphs.p2'
                    )}
                  </p>
                </>
              }
            />
          </>
        ),
      };
    case 'settings':
      return {
        title: 'settings',
        template: (
          <Settings
            onClose={callback}
            onLocateChange={onLocateChange}
            locale={locale}
          />
        ),
      };
    case 'contribute':
      return {
        title: 'contribute',
        template: (
          <>
            <p>
              Do you have any NPI source you would want to share with us? Please
              follow this link:
            </p>
            <a
              className="ld-link"
              target="_blank"
              rel="noopener noreferrer"
              href="https://docs.google.com/forms/d/e/1FAIpQLSfDWe2qlzUnd3e-YYspMzT9adUswDEYIdJMb7jz7ule34-yiA/viewform"
            >
              Submit data
            </a>
            <p>
              Additionally, you can become part of the project by joining one of
              the following teams:
              <ul>
                <li>Editors</li>
                <li>Devs</li>
                <li>UI/UX</li>
              </ul>
            </p>
            <p>
              Contact us via Telegram at:
              <br />
              <a
                className="ld-link"
                target="_blank"
                rel="noopener noreferrer"
                href="https://t.me/ProjectLockdown"
              >
                {' '}
                @ProjectLockdown
              </a>
            </p>
          </>
        ),
      };

    case 'updates':
      return {
        title: 'updates',
        template: <Ticker />,
      };

    default:
      return {
        title: '',
        template: <></>,
      };
  }
};

export class TabMenu extends Component {
  constructor(props) {
    super(props);
    this.state = {
      activeItem: 'info',
      showLateralMenu: false,
      showMenu: false,
      currentDropdown: 1,
    };
    this.showSideBar = this.showSideBar.bind(this);
    this.closeNavbar = this.closeNavbar.bind(this);
    this.switchContent = this.switchContent.bind(this);
    this.onDropDown = this.onDropDown.bind(this);
  }

  componentDidMount() {
    let i = 0;

    installMediaQueryWatcher(`(min-width: 960px)`, matches => {
      this.setState({
        isMobile: !matches,
      });
      if (matches && i > 0) {
        // This is super ugly, but this fires on pageload and causes the focus to be set on the menu :/
        i++;
        this.props.close();
      }
    });
  }

  showSideBar() {
    this.setState({
      showLateralMenu: true,
      showMenu: true,
    });
  }
  switchContent(val) {
    if (val === 'settings' && this.state.updateAvailable) {
      this.setState({
        updateAvailable: false,
      });
    }

    if (this.state.isMobile && this.props.opened && val === this.prevVal) {
      this.props.close();
      this.setState({
        activeItem: this.prevVal,
      });
      this.prevVal = '';
      return;
    }

    // this.props.changeRoute(renderMenu(val));

    this.prevVal = val;
    this.setState({
      activeItem: val,
      showLateralMenu: val === this.state.activeItem ? false : true,
    });
  }

  closeNavbar() {
    this.setState({
      showLateralMenu: false,
      showSideBar: false,
      activeItem: 'info',
    });
  }

  onDropDown(id) {
    this.setState({
      currentDropdown: id,
    });
  }

  render() {
    const { activeItem, currentDropdown } = this.state;
    return this.state.showLateralMenu || this.props.isMobile === true ? (
      <>
        <div className="menu-overlay"></div>
        <main id="main" className="ld-menu">
          <div className="ld-menu-nav">
            <button className="menu-close-btn" onClick={this.closeNavbar}>
              {closeIcon}
            </button>
            <nav>
              <Tabs
                onClose={this.closeNavbar}
                switchContent={this.switchContent}
              >
                <button id="info">info</button>
                <button id="settings">settings</button>
                <button id="updates">updates</button>
                <button id="contribute">contribute</button>
              </Tabs>
            </nav>
          </div>
          <div className={this.state.showLateralMenu ? 'ld-menu--content' : ''}>
            <div className="mb-only">
              <div className="ld-menu--header">
                <h1>{activeItem}</h1>
              </div>
            </div>
            {
              renderMenu(
                activeItem,
                this.closeNavbar,
                currentDropdown,
                this.onDropDown,
                this.props.onLocateChange,
                this.props.locale
              ).template
            }
          </div>
        </main>
      </>
    ) : (
      <button onClick={this.showSideBar} className="menu-side-btn">
        {burger}
      </button>
    );
  }
}
