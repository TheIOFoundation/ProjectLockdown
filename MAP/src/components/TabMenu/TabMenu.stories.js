import { TabMenu } from './TabMenu';
import * as ExpandableStories from '../Expandable/Expandable.stories';
import * as SettingsStories from '../Settings/Settings.stories';
import * as TabsStories from '../Tabs/Tabs.stories';
import * as TickerStories from '../Ticker/Ticker.stories';

const parameters = {
  title: 'Component/TabMenu',
  component: TabMenu,
};

const Template = args => <TabMenu {...args} />;

export const Primary = Template.bind({});
Primary.args = {
  ...ExpandableStories.Primary.args,
  ...SettingsStories.Primary.args,
  ...TabsStories.Primary.args,
  ...TickerStories.Primary.args,
  opened: false,
  close: () => {
    console.log('close');
  },
  onLocateChange: a => {
    console.log('onLocateChange arg:', a);
  },
  isMobile: false,
};

export default parameters;
