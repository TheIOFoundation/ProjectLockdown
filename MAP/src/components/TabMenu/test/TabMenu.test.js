import React from 'react';
import renderer from 'react-test-renderer';
import { TabMenu } from '../TabMenu';
const props = {
    dark: "false",
    setDarkMode: jest.fn()
}

window.matchMedia = window.matchMedia || function() {
    return {
        matches: true,
        addEventListener: jest.fn(),
    };
};
it('renders correctly when there are no items', () => {
    const tree = renderer.create(<TabMenu {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });