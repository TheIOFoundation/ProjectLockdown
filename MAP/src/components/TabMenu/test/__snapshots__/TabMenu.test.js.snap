// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly when there are no items 1`] = `
<button
  className="menu-side-btn"
  onClick={[Function]}
>
  <svg
    fill="currentColor"
    height="14"
    stroke="currentColor"
    viewBox="0 0 22 14"
    width="22"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      clipRule="evenodd"
      d="M0.5 2.33333V0H21.5V2.33333H0.5ZM0.5 8.16667H21.5V5.83333H0.5V8.16667ZM0.5 14H21.5V11.6667H0.5V14Z"
      fill="currentColor"
      fillOpacity="1"
      fillRule="evenodd"
    />
  </svg>
</button>
`;
